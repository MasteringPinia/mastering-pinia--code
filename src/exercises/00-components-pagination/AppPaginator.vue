<script lang="ts" setup>
import { computed } from 'vue'

const props = withDefaults(
  defineProps<{
    currentPage: number
    perPage: number
    total: number
  }>(),
  { perPage: 10 },
)

const emit = defineEmits<{
  (event: 'update:currentPage', page: number): void
}>()
// const emit = defineEmits({
//   'update:currentPage': (page: number) => typeof page === 'number',
// })
const totalPages = computed(() => {
  return Math.ceil(props.total / props.perPage) || 1
})

function previousPage() {
  if (props.currentPage > 1) emit('update:currentPage', props.currentPage - 1)
}

function nextPage() {
  if (props.currentPage < totalPages.value) emit('update:currentPage', props.currentPage + 1)
}
</script>

<template>
  <div class="flex items-center justify-center">
    <button data-test="previous" class="button" :disabled="currentPage < 2" @click="previousPage">Previous</button>
    <span data-test="text" class="mx-3">{{ currentPage }} / {{ totalPages }}</span>
    <button data-test="next" :disabled="currentPage >= totalPages" @click="nextPage">Next</button>
  </div>
</template>
