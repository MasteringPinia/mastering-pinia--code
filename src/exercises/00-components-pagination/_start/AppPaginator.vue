<script lang="ts">
import { computed, defineComponent } from 'vue'

export default defineComponent({
  model: {
    prop: 'currentPage',
    event: 'update:currentPage',
  },

  props: {
    currentPage: {
      type: Number,
      required: true,
    },
    perPage: {
      type: Number,
      default: 10,
    },
    total: {
      type: Number,
      required: true,
    },
  },

  emits: {
    'update:currentPage': (page: number) => typeof page === 'number',
  },

  // a ne pas faire
  // setup({ total }, { emit }) {
  setup(props, { emit }) {
    return {}
  },
})
</script>

<template>
  <div class="flex items-center justify-center">
    <button data-test="previous" class="button">Previous</button>
    <span data-test="text" class="mx-3">0 / 35</span>
    <button data-test="next">Next</button>
  </div>
</template>
