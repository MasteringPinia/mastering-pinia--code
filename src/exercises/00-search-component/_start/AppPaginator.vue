<script setup lang="ts"></script>

<template>
  <div class="flex items-center justify-center">
    <button data-test="previous" class="button">Previous</button>
    <span data-test="text" class="mx-3">0 / 35</span>
    <button data-test="next">Next</button>
  </div>
</template>
