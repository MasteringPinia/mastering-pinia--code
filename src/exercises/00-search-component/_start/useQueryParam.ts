// import {
//   // Some interesting types
//   LocationQueryRaw,
//   LocationQueryValue,
//   LocationQueryValueRaw,
// } from 'vue-router'

import { ref } from 'vue'

export function useQueryParam() {
  // This is probably something else
  return ref()
}
