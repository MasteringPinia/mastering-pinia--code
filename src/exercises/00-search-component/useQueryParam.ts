import { computed, nextTick } from 'vue'
import { LocationQueryRaw, LocationQueryValue, LocationQueryValueRaw, useRoute, useRouter } from 'vue-router/auto'

let pendingQuery: LocationQueryRaw | null = null

export function useQueryParam<T = string>(
  key: string,
  {
    parseValue = (value: LocationQueryValue | LocationQueryValue[]): T =>
      // @ts-expect-error: '' is not assignable to T, but in those scenarios, the parseValue should be overridden
      Array.isArray(value) || !value ? '' : value,
    serializeValue = (value: T): LocationQueryValueRaw | LocationQueryValueRaw[] => String(value),
    deleteIf = (value: T) => !value,
  } = {},
) {
  const route = useRoute()
  const router = useRouter()

  return computed<T>({
    get: () => parseValue(route.query[key]),
    set: value => {
      const query = pendingQuery || { ...route.query }
      query[key] = serializeValue(value)
      if (deleteIf(value)) {
        delete query[key]
      }

      if (!pendingQuery) {
        nextTick().then(() => {
          router.push({ query })
          pendingQuery = null
        })
        pendingQuery = query
      }
    },
  })
}
