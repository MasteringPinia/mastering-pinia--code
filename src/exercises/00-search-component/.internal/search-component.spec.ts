import { mockHttpRequests, delay } from '@tests/mocks/server'
import TestComponent from '../index.vue'
import { disableAutoUnmount, enableAutoUnmount, mount } from '@vue/test-utils'
import { getRouter } from 'vue-router-mock'
import { useQueryParam } from '../useQueryParam'
import AppPaginator from '../AppPaginator.vue'
import { describe, it, expect, beforeEach, afterEach, afterAll, vi, SpyInstance } from 'vitest'

// TODO: add other tests for the instructions

describe('Search Component', () => {
  let router = getRouter()
  let pushSpy = router.push as SpyInstance

  mockHttpRequests()

  enableAutoUnmount(afterEach)
  afterAll(disableAutoUnmount)

  beforeEach(async () => {
    router = getRouter()
    pushSpy = router.push as SpyInstance
    router.reset()
    await router.push('/')
    vi.clearAllMocks()
  })

  afterAll(async () => {
    await router.getPendingNavigation()
    vi.restoreAllMocks()
  })

  it('can change the page', async () => {
    const wrapper = mount(TestComponent, { attachTo: document.body })
    await wrapper.get('form input').setValue('sales')
    await wrapper.get('form').trigger('submit')
    await delay(20)
    await wrapper.get('button[data-test=next]').trigger('click')
    await router.getPendingNavigation()
    expect(wrapper.get('[data-test=text]').text()).toBe('2 / 5')
  })

  it('computes the total number of pages', async () => {
    const wrapper = mount(AppPaginator, {
      props: {
        currentPage: 1,
        perPage: 10,
        total: 50,
      },
    })

    expect(wrapper.get('[data-test=text]').text()).toBe('1 / 5')
  })

  it('disables the previous button on the first page', async () => {
    const wrapper = mount(AppPaginator, {
      props: {
        currentPage: 1,
        perPage: 10,
        total: 50,
      },
    })

    expect(wrapper.get('[data-test=previous]').attributes('disabled')).toBeDefined()
    expect(wrapper.get('[data-test=next]').attributes('disabled')).toBeUndefined()
  })

  it('disables the next button on the last page', async () => {
    const wrapper = mount(AppPaginator, {
      props: {
        currentPage: 5,
        perPage: 10,
        total: 50,
      },
    })

    expect(wrapper.get('[data-test=next]').attributes('disabled')).toBeDefined()
    expect(wrapper.get('[data-test=previous]').attributes('disabled')).toBeUndefined()
  })

  it('handles the amount of pages correctly', async () => {
    const wrapper = mount(AppPaginator, {
      props: {
        currentPage: 1,
        perPage: 10,
        total: 51,
      },
    })

    expect(wrapper.get('[data-test=text]').text()).toBe('1 / 6')
  })

  it('triggers changes of the page when pressing next', async () => {
    const wrapper = mount(AppPaginator, {
      props: {
        currentPage: 1,
        perPage: 10,
        total: 51,
      },
    })

    await wrapper.get('[data-test=next]').trigger('click')
    expect(wrapper.emitted('update:currentPage')).toEqual([[2]])
  })

  it('show results when using the form', async () => {
    const wrapper = mount(TestComponent)
    await wrapper.get('form input').setValue('sales')
    await wrapper.get('form').trigger('submit')
    await delay(20)

    const contacts = wrapper.findAll('ul li')
    expect(contacts.length).toBe(12)
    contacts.forEach((contact, i) => {
      expect(contact.text().trim()).toMatchSnapshot(`contact i${i}`)
    })
  })

  it.todo('shows an empty results message')
  it.todo('can customize with a slot #no-results the empty results message')
  it.todo('shows an error message')
  it.todo('can customize with a slot #error the error message')
  it.todo('customize the results of SearchPaginator')
  it.todo('displays a loading state while fetching')
  it.todo('waits 200ms before displaying the loading state')

  // describe('router', () => {

  it('uses the searchText query for the search', async () => {
    await router.setQuery({ searchText: 'john' })
    const wrapper = mount(TestComponent)
    expect(wrapper.get<HTMLInputElement>('form input').element.value).toBe('john')
  })

  it.todo('uses the page query for the search')

  it('updates the input with searchQuery when navigating', async () => {
    const wrapper = mount(TestComponent)
    await router.setQuery({ searchText: 'john' })
    expect(router.currentRoute.value.query.searchText).toBe('john')
    expect(wrapper.get<HTMLInputElement>('form input').element.value).toBe('john')
  })

  it('saves the search text on the url', async () => {
    const wrapper = mount(TestComponent)
    // expect(pushSpy.callCount).toBe(1)
    await wrapper.get('form input').setValue('jason')
    await wrapper.get('form').trigger('submit')
    await router.getPendingNavigation()
    // expect(pushSpy.callCount).toBe(2)
    expect(router.currentRoute.value.query).toMatchObject({
      searchText: 'jason',
    })

    it('calls router.push() once when changing multiple query params', async () => {
      pushSpy.mockClear()
      const wrapper = mount({
        setup() {
          const a = useQueryParam('a')
          const b = useQueryParam('b')

          return { a, b }
        },
        render: () => null,
      })

      expect(pushSpy).toHaveBeenCalledTimes(0)

      wrapper.vm.a = 'c'
      wrapper.vm.b = 'c'
      expect(pushSpy).toHaveBeenCalledTimes(0)

      await delay(20)
      expect(pushSpy).toHaveBeenCalledTimes(1)
    })
  })

  // })
})
