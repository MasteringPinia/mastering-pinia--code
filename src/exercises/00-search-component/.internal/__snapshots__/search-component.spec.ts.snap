// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Component > show results when using the form > contact i0 1`] = `"Jason Sales MasterVP Sales"`;

exports[`Search Component > show results when using the form > contact i1 1`] = `"Rosina CastanoSales Representative"`;

exports[`Search Component > show results when using the form > contact i2 1`] = `"Karilynn ShropsheirSales Associate"`;

exports[`Search Component > show results when using the form > contact i3 1`] = `"Sharla HollowsDirector of Sales"`;

exports[`Search Component > show results when using the form > contact i4 1`] = `"Andeee LibbisSales Associate"`;

exports[`Search Component > show results when using the form > contact i5 1`] = `"Sondra ScotchmurDirector of Sales"`;

exports[`Search Component > show results when using the form > contact i6 1`] = `"Kellia JayeVP Sales"`;

exports[`Search Component > show results when using the form > contact i7 1`] = `"Rosetta MagovernSales Associate"`;

exports[`Search Component > show results when using the form > contact i8 1`] = `"Torr RabierSales Associate"`;

exports[`Search Component > show results when using the form > contact i9 1`] = `"Barbra WileSales Representative"`;

exports[`Search Component > show results when using the form > contact i10 1`] = `"Tucky KirkmanSenior Sales Associate"`;

exports[`Search Component > show results when using the form > contact i11 1`] = `"Emma DumingoSenior Sales Associate"`;
